[28기 김정현] [Refactor] conflict 일괄 수정

KimJeongHyun opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [ ] 기능 추가
- [x] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
 - 공통 컴포넌트인 Navigation bar 작성
 - 스크롤 다운 hide, 스크롤 업 show animation 구현되어야함.
 - 버튼 클릭시 알맞는 페이지로 Routing 진행되어야함
 - 햄버거 버튼 클릭시 hidden menu showing 되어야함 

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 전체를 Wrapping하는 Nav
- Router.js를 묶는 Wrapper로서, 어떤 페이지든 Navbar가 포함됨.
- Navbar는 반응형으로 width가 변경됨.

2. 버튼 클릭시 라우팅 
 - 아이콘 클릭시 그에 맞는 동작을 할 수 있도록 a 태그로 Wrapping

3. 스크롤 이벤트
- 스크롤 업 이벤트 발생시 show, 스크롤 다운 이벤트 발생시 hide 되도록 적용.

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
 - 닥터마틴 사이트에서 적용되고 있는 보이지 않는 margin을 표현한 flex 축약형 속성에 대해 생각하고 있습니다.
 - 포지션 관리에 대해 집중하고 있습니다.

<br />

## :: 기타 질문 및 특이 사항
- 햄버거 버튼 클릭시 hidden menu가 점멸하는 부분은 구현 예정이나 Main 작업으로 인해 잠깐 보류 상태입니다.

KimJeongHyun committed:
[Add] Navbar 작성 완료

KimJeongHyun committed:
[Modify] Navigation component 위치 변경
 - Nav bar가 모든 page에 적용되는 바 router의 wrapper로 적용
 - 관련 태그 및 스타일 Router.js 및 index.scss로 이관

saengmotmi commented on src/components/Nav/Nav.js:
```diff
@@ -1,3 +1,48 @@
+
 export default function Nav() {
-  return <div className="navbar" />;
+  return (
+    <header className="navHead">
+      <section className="navUpper">
+        <div className="navUpperFixedWidth">
+          <nav>
+            <a href="#!">로그인</a>
+            <a href="#!" className="strong">
```
nav가 header 안쪽에 들어가 있는게 어색해보여요 제 생각에는 nav가 가장 바깥쪽에 쓰여야 하지 않을까 싶네요

KimJeongHyun replied:
실제로도 레이아웃 클로닝하면서 어색함을 느꼈던 부분입니다. 아예 nav대신 div tag를 적용했습니다.

saengmotmi commented on src/Router.js:
```diff
@@ -1,13 +1,17 @@
-        <Route path="/" element={<Main />} />
-      </Routes>
-    </BrowserRouter>
+    <main className="headContainer">
```
이 부분 역시 Main 컴포넌트 안쪽으로 들어가는게 맞지 않을까 싶어요 main 태그는 페이지에서 사용되는 메인 컨텐츠를 포함하는 태그인데 (ex. 인스타그램 피드) 지금은 앱 전체를 감싸고 있네요

KimJeongHyun replied:
@saengmotmi 

이 부분이 어떤 의미인지 이해되지 않았습니다 ㅠ.ㅠ

우선 제 의도는 

사이트가 현재 랜딩 페이지를 제외하곤 'width:1140px'를 기준으로 정렬되고 있으며
Nav bar 또한 모든 페이지에 포함됩니다.
따라서 전체를 Wrapping하는 headContainer에 min-width:1140으로 명시하고 Nav를 안에 포함시킨 것인데,
main 태그가 아닌, 예를 들면 div 태그로 변경하는 방향으로 하면 될까요??

saengmotmi replied:
Nav 컴포넌트는 Main과 무관하게 가로 100%를 차지하도록 하면 되고
main 태그를 Main 컴포넌트 안쪽으로 넣는게 맞지 않을까 싶어서 말씀드린거였어요

그리고 만약 브라우저 너비가 1140 이상보다 커지게 되면 nav 영역이 그 이상 늘어나지 않아 어색하지 않을까요?

KimJeongHyun replied:
main 태그는 의도한 것과 다른 의미를 내포하므로 div로 변경했습니다.

Nav 포함, 대부분의 이미지는 1140px를 기준으로 실제 사이트에서 렌더링되고 있기는 하지만
개선 사항으로 판단하여 브라우저의 크기 변화에 따라 nav 영역이 확장되거나 줄어들거나 하도록 적용했습니다.

추가로 변경에 따라 input 태그의 위치가 어그러져 그 앞단에 보이지 않는 간격을 flex로 준 것과 같이
flex:auto 속성 적용으로 자연스럽게 배치되도록 변경했습니다.

saengmotmi commented on src/components/Nav/Nav.scss:
```diff
@@ -0,0 +1,107 @@
+.navHead {
+  box-shadow: 0 3px 7px 0 rgb(0 0 0 / 10%);
+  .navUpper {
+    height: 40px;
```
```suggestion
.navHead {
  box-shadow: 0 3px 7px 0 rgb(0 0 0 / 10%);
  
  .navUpper {
    height: 40px;
```
새로운 셀렉터 나올 때마다 한줄 씩 띄워주세요

KimJeongHyun replied:
확인하여 적용되었습니다!

saengmotmi commented on src/components/Nav/Nav.scss:
```diff
@@ -0,0 +1,107 @@
+    background-color: #fafafa;
+    .navUpperFixedWidth {
+      position: relative;
+      width: $widthSize;
```
widthSize라는 상수 변수가 어색해보이는데 다음 두가지 이유로 그렇습니다
- '가로 길이' 라는 네이밍 자체로 알 수 있는 정보가 없음
- 변수로 지정했다는 건 그 자체로 의미가 있는 값이거나 재사용될 값이라는 뜻일텐데 이 부분 또한 네이밍만 보고서는 불명확함

KH2Wone replied:
`$mainWidthSize`, `$mainContentWidth` 정도로 생각해보았는데 이런 네이밍이 괜찮을까요?

saengmotmi replied:
@KH2Wone 넵 더 좋은 이름이 있을지는 고민해봐야겠지만 이런 방향이 되어야 할거 같습니다

KimJeongHyun replied:
희원님께서 변경하신 이력에 맞게 변수명 재정립 진행했습니다!

saengmotmi commented on src/components/Nav/Nav.scss:
```diff
@@ -0,0 +1,107 @@
+        justify-content: flex-end;
+        line-height: 40px;
+        a {
+          font-size: $smallFontSize12;
+          color: $colorBlack111;
```
```suggestion
          font-size: $fontSize12;
          color: $colorBlack111;
```
- smallFontSize -> 그냥 fontSize 해도 되지 않을까요?
- black은 당연히 color이므로 black 그냥 쓰시면 될거 같고 111의 네이밍 이유도 궁금해요 어떤 의도인가요?

KH2Wone replied:
안녕하세요, 종택님!
- `smallFontSize`는 폰트 사이즈가 여러 개라서 `title`, `medium`, `default`, `small`로 크게 나누었습니다.
그리고 그 안에 포함된 항목들이 사이즈만 다르고 특징이 없어서 `px` 사이즈를 변수 끝에 기재하였습니다.
아니면 이 또한 명확하지 않으므로 종택님 말씀대로 `fontSize64`, `fontSize14`로 수정하는 것도 생각해 보았는데 괜찮을까요?
- `color`같은 경우에는 닥터마틴 사이트의 경우 유사한 black이어도 hex가 다른 경우가 있어 뒤에 hex를 작성하였습니다.
`color`라는 이름을 제외하고 `$black111`로 하는 것도 네이밍에 문제가 없을지 여쭤보고 싶습니다.

saengmotmi replied:
@KH2Wone 
- 그런 목적이라면 차라리 font-small, font-sm, small, sm, font-title 등의 네이밍이 더 좋을 것 같습니다. 구체적인 픽셀이 들어간다면 네이밍을 해주는 의미가 많이 퇴색되는거 같아서요
- color 역시 구체적인 hex 값을 적어주는 것보다 light-gray, gray, deep-gray 등의 이름으로 구분해주거나 아니면 gray100, gray200, gray300 등 진하기를 숫자로 표현하는 방식이 나아보여요

KimJeongHyun replied:
희원님께서 변경하신 이력에 맞게 변수명 재정립 진행했습니다!

saengmotmi commented on src/components/Nav/Nav.scss:
```diff
@@ -0,0 +1,107 @@
+            margin-left: 20px;
+          }
+        }
+        a.strong {
+          font-weight: 700;
+        }
```
.strong 같은 클래스명은 common 같은 곳에 빼놓고 className="strong" 하면 어디서든 700 먹을수 있도록 하면 좋겠네요

KimJeongHyun replied:
프론트엔드 분들과의 논의 하에 common.scss 에 *.strong을 통해 불러오도록 반영했습니다!

saengmotmi commented on src/components/Nav/Nav.scss:
```diff
@@ -0,0 +1,107 @@
+    }
+  }
+  .navBottom {
+    width: 100%;
```
이미 가로 100% 아닐까요?

KimJeongHyun replied:
지적하신 것과 마찬가지로 불필요한 100% 값을 일괄 삭제했습니다.

saengmotmi commented on src/components/Nav/Nav.scss:
```diff
@@ -0,0 +1,107 @@
+    margin: 10px auto;
+    .navBottomFixedWidth {
+      display: flex;
+      flex-direction: row;
```
row도 flex 컨테이너에서 기본값입니다

KimJeongHyun replied:
항상 명시했던 것이라 기본값으로 설정되어 있을 줄 몰랐네요.... 해당 건 포함하여 모두 삭제했습니다.

saengmotmi commented on src/components/Nav/Nav.scss:
```diff
@@ -0,0 +1,107 @@
+      flex-direction: row;
+      align-items: center;
+      position: relative;
+      width: 1140px;
```
만약 화면 크기가 1140 픽셀 보다 작아지면 어떻게 될까요?

KimJeongHyun replied:
@saengmotmi 

레퍼런스인 닥터마틴에서는 작아지는 경우에 대해 따로 처리해둔 것이 없었습니다.

max-width:1140px
width:100% 

로 다음 커밋에 반영하겠습니다.

saengmotmi reviewed (commented):
리뷰 확인해주세요!

saengmotmi commented:
리뷰 요청하실 때는 리뷰 요청 라벨 달아주세요!

KimJeongHyun committed:
[Add] Main 1, 2문단 및 custom scroll hook 작성
 - 1문단 이미지 슬라이더 완료
 - 2문단 베스트 아이콘 레이아웃 및 슬라이더 완료
 - 커스텀 scroll fadein 활용을 통한 무한 슬라이더 구현
 - Main.js 컴포넌트 분리 진행
 - 전역 스크롤 이벤트 state 설정을 통해 과한 event 발생 방지(확인필요)

KimJeongHyun committed:
[Refactor] scss 리팩토링
- 변수 이름 재정립
- nav 태그 정리
- scss 선택자 한 줄씩 띄우도록 정리
- strong 선택자 common으로 이관
- 기본 값 기재된 것들 삭제

KimJeongHyun committed:
[Fix] main branch 병합

KimJeongHyun committed:
[Modify] main merge 후 미반영 사항 정리
-index.scss에  headContainer style 반영

KimJeongHyun committed:
[Fix] 오 커밋된 페이지 롤백
- Main 부분 파일 삭제

KimJeongHyun committed:
[Refactor] Nav.scss 리팩토링 작업 및 Router.js 태그 수정
- Router.js 감싸는 태그  main에서 div로 변경
- Nav.js 불필요한 100% 값 일괄 삭제
- 공통 style mixin으로 통합
- 반응형으로 수정

saengmotmi commented on package.json:
```diff
@@ -11,6 +11,7 @@
     "@testing-library/jest-dom": "^5.16.1",
     "@testing-library/react": "^12.1.2",
     "@testing-library/user-event": "^13.5.0",
+    "lodash": "^4.17.21",
```
사용 필요한가요? 불필요하면 삭제해주세요

KimJeongHyun replied:
throttle 적용을 깜빡해두었습니다 ^^;;; 다시 적용한 상태로 올렸습니다.

saengmotmi replied:
@KimJeongHyun 단순히 쓰로틀 사용을 위해서라면 설치하지 않는게 어떨까요. 검색해보시면 직접 구현된 throttle을 사용해볼 수도 있습니다.

saengmotmi commented on src/Router.js:
```diff
@@ -1,13 +1,32 @@
+
+  useEffect(() => {
+    function handleScroll() {
+      windowY > window.scrollY ? setVisibleValue(true) : setVisibleValue(false);
```
```suggestion
      sthBlahBlah ? setVisibleValue(true) : setVisibleValue(false);
```
조건식 그대로 두기보다 좋은 변수명 있다면 이름 지어주는게 좋을 것 같아요

KimJeongHyun replied:
저번에 예시들어주신 것처럼 isScrollUp 이라는 boolean 변수로 제어하도록 만들었습니다.

saengmotmi commented on src/Router.js:
```diff
@@ -1,13 +1,32 @@
+  useEffect(() => {
+    function handleScroll() {
+      windowY > window.scrollY ? setVisibleValue(true) : setVisibleValue(false);
+      setWindowY(window.scrollY);
```
위에서 visible 상태를 관리하고 있는데 scrollY도 저장할 필요가 있나요?

KimJeongHyun replied:
state의 비동기성을 통해

handleScroll에서 호출하는 windowY는 현재 바라보고 있는 window.scrollY보다 작거나, 큰 값으로 결정되어

스크롤 up 또는 down 상태를 결정하게끔 되어있습니다.

혹시 더 좋은 방법이 있을까요??

saengmotmi replied:
앱 전체에서 유일하고 router에서만 신경쓰면 되는 값이기 때문에, 그리고 해당 값의 업데이트가 setVisibleValue에 종속되기 때문에 컴포넌트 바깥에 let으로 선언하는 값으로 사용하던가 렌더링과 무관한 useRef를 사용하는 것이 어떨까 싶습니다

saengmotmi replied:
어차피 보이고 안보이고는 setVisibleValue로 컨트롤하니까요

saengmotmi commented on src/Router.js:
```diff
@@ -1,13 +1,32 @@
-        <Route path="/" element={<Main />} />
-      </Routes>
-    </BrowserRouter>
+    <div className="headContainer">
+      <Nav visibleValue={visibleValue} />
+      <BrowserRouter>
+        <Routes>
+          <Route path="/" element={<Main />} />
+        </Routes>
+      </BrowserRouter>
+    </div>
```
```suggestion      
      <BrowserRouter>
        <Nav visibleValue={visibleValue} />
        <Routes>
          <Route path="/" element={<Main />} />
        </Routes>
      </BrowserRouter>
```
제가 제안드렸던건 이러한 형태였는데 확인 부탁드려요

KimJeongHyun replied:
확인해보니 현재 작업 중인 Main branch에서도 headContainer의 유무가 크게 중요하지 않은 것 같습니다.

따라서 현재 Nav branch에서는 삭제했고 컴포넌트 위치 또한 제안해주신 것과 같이 변경되었습니다.

saengmotmi commented on src/components/Nav/Nav.scss:
```diff
@@ -0,0 +1,146 @@
+  }
+}
+
+.navHead.hide {
+  opacity: 0;
+}
```
```suggestion
.navHead {
  &.hide {
    opacity: 0;
  }
}
```
동일한 내용을 위와 같이 표현할 수 있습니다

KimJeongHyun replied:
확인하여 적용되었습니다.

saengmotmi commented on src/hooks/useScroll.js:
```diff
@@ -0,0 +1,26 @@
+import { useEffect, useRef, useCallback } from 'react';
+
+export function useScrollFadeIn() {
```
굿굿 👍

saengmotmi commented on src/styles/common.scss:
```diff
@@ -1,12 +1,16 @@
   clear: both;
 }
+
+*.strong {
```
```suggestion
.strong {
```
전체 선택자 없이도 사용 가능하지 않을까요?

KimJeongHyun replied:
확인하여 적용되었습니다.

saengmotmi reviewed (commented):
리뷰 반영해주세요! 그리고 리뷰 수정하셨으면 기존에 달려 있던 수정 요청 라벨은 떼주시면 됩니다
+ PR 올리실 때 이름도 표기해주세요

KimJeongHyun committed:
[Refactor] scroll throttle 적용, headContainer 삭제 등
- nav scroll에 throttle 적용.
- headContainer 삭제 (불필요한 것으로 판단함)
- Nav.scss 선택자 정리
- index.scss에서 headContainer 스타일 삭제
- common strong 선택자 정리

saengmotmi commented:
@KimJeongHyun 충돌 해결해주세요!

saengmotmi reviewed (commented):
마이너한 부분이라 코멘트만 남겨두겠습니다 고생하셨어요 🔥

KimJeongHyun committed:
[Refactor] conflict 일괄 수정

KimJeongHyun commented:
@saengmotmi 

커스텀 훅 썼던 것처럼 throttle도 커스텀으로 따로 빼어 구현해보도록 하겠습니다.
그리고 다시보니 코멘트 주신 것처럼 따로 ref로 빼어 동작하는 것이 더 효율적인 동작이 될 수 있을 것 같습니다. 감사합니다.

conflict 해결되었습니다!

saengmotmi commented:
고생하셨습니다! 🔥
